File: 0985-bag-of-tokens/0985-bag-of-tokens.cpp
class Solution {
public:
    int bagOfTokensScore(vector<int>& tokens, int power) {
        sort(tokens.begin(),tokens.end()); 
        int left = 0; 
        int right = tokens.size()-1; 
        int maxScore = 0; 
        int currScore = 0; 

        while( left<= right){
            if( power >= tokens[left]){
                power-=tokens[left++];
                currScore+=1;
                maxScore = max(maxScore,currScore);
            }
            else if( currScore>=1){
                power += tokens[right--]; 
                currScore-=1;
            }
            else break;
        }
        return maxScore;
        
    }
};

File: 0985-bag-of-tokens/0985-bag-of-tokens.java
import java.util.Arrays;

class Solution {
    public int bagOfTokensScore(int[] tokens, int P) {
        Arrays.sort(tokens);
        int currScore = 0;
        int maxScore = 0;
        int l = 0, r = tokens.length - 1;

        // GREEDY
        while (l <= r) {
            if (P >= tokens[l]) {
                P -= tokens[l];
                currScore++;
                maxScore = Math.max(maxScore, currScore);
                l++;
            } else if (currScore >= 1) {
                P += tokens[r];
                currScore--;
                r--;
            } else {
                return maxScore;
            }
        }

        return maxScore;
    }
}
